### ISSUE TITLE
pgcompat: clarify our PostgreSQL compatibility story

### ISSUE THREAD

**Issue by benesch** [MaterializeInc/materialize#8086]
Our current position is that our wire protocol is 100% PostgreSQL compatible, while our SQL dialect is PostgreSQL compatible "when possible." This issue tracks two work items:

 - [ ] Adding a description of the current situation in the docs. (As far as I know, we don't currently state our compatibility policy explicitly anywhere externally.)

 - [ ] Nailing down what "when possible" means. 

    A good example of this came up recently in a [conversation in our community Slack](https://materializecommunity.slack.com/archives/C015KDVS7EV/p1629659994133100), where a user was surprised that our regex syntax did not match PostgreSQL's. There's a good reason for this: PostgreSQL has a custom regex implementation that we can't easily steal. 

    What do we want to do in cases like this? A PostgreSQL-compatible regex engine is easily weeks to a month of work plus the permanent maintenance burden of fixing pgcompat bugs, whereas the Rust `regex` crate was integrated in less than a day and there is no ongoing maintenance burden. But at the same time, using the Rust `regex` crate imposes a permanent burden on our users, who need to rewrite their regexes that they used in PostgreSQL for Materialize. I can imagine that some of the differences in dialects result in subtle incorrectness, which is my least favorite kind of bug.

    In particular it'd be nice to have a rule of thumb to apply to help make decisions like these. Whatever we decide, we should document it, and also start accumulating a list of known deviations from Postgres.

cc @TonyRippy @mjibson

**Comment by mjibson** [comment]
There are also things we will never support due to dataflow (`now() in materialized views`), which should also be documented as a separate category.

**Comment by jseldess** [comment]
I'm adding `PostgreSQL compatibility` as a planned page in the new docs information architecture: https://github.com/MaterializeInc/materialize/issues/18508#issuecomment-1490626224. 

I'm thinking we can add a reference page something like what [CockroachDB's](https://www.cockroachlabs.com/docs/stable/postgresql-compatibility.html) that explains we are wire-compatible and lists which parts of the PostgreSQL dialect we don't support and which parts we support but with different behavior. Does that sound right? 

CockroachDB also has a more general reference page listing [SQL feature support](https://www.cockroachlabs.com/docs/stable/sql-feature-support.html), and we could consider something like that. But a bigger question here is how much we want to focus on things that are only really related to native Materialize tables vs working with data ingested into Materialize (querying, RBAC, etc.).

**Comment by mjibson** [comment]
I agree we should document anything that has different behavior but is typed the same (floats come to mind for example, where under the hood we convert them to i128s instead of doing spec float math on them).
